diff --git a/include/android_stub/android/hardware_buffer.h b/include/android_stub/android/hardware_buffer.h
index aa03b4b096e..7a51ede4736 100644
--- a/include/android_stub/android/hardware_buffer.h
+++ b/include/android_stub/android/hardware_buffer.h
@@ -51,6 +51,10 @@
 
 #include <android/rect.h>
 
+#if !defined(__BIONIC__) && !defined(__INTRODUCED_IN)
+#define __INTRODUCED_IN(x)
+#endif
+
 __BEGIN_DECLS
 
 /**
diff --git a/src/gfxstream/guest/android/GrallocEmulated.cpp b/src/gfxstream/guest/android/GrallocEmulated.cpp
index 66dadbdfbeb..20253b595f1 100644
--- a/src/gfxstream/guest/android/GrallocEmulated.cpp
+++ b/src/gfxstream/guest/android/GrallocEmulated.cpp
@@ -18,6 +18,7 @@ static constexpr int numFds = 0;
 static constexpr int numInts = 1;
 
 #define DRM_FORMAT_R8_BLOB fourcc_code('9', '9', '9', '9')
+#define DRM_FORMAT_DEPTH32_STENCIL8 fourcc_code('D', 'F', 'S', '8')
 
 template <typename T, typename N>
 T DivideRoundUp(T n, N divisor) {
@@ -76,6 +77,8 @@ std::optional<uint32_t> AhbToDrmFormat(uint32_t ahbFormat) {
             return DRM_FORMAT_ABGR2101010;
         case GFXSTREAM_AHB_FORMAT_Y8Cb8Cr8_420:
             return DRM_FORMAT_NV12;
+        case GFXSTREAM_AHB_FORMAT_D32_FLOAT_S8_UINT:
+            return DRM_FORMAT_DEPTH32_STENCIL8;
     }
     return std::nullopt;
 }
@@ -270,6 +273,22 @@ const std::unordered_map<uint32_t, DrmFormatInfo>& GetDrmFormatInfoMap() {
                      },
                  },
          }},
+        {DRM_FORMAT_DEPTH32_STENCIL8,
+         {
+             .androidFormat = GFXSTREAM_AHB_FORMAT_D32_FLOAT_S8_UINT,
+             .virglFormat = VIRGL_FORMAT_Z32_FLOAT_S8X24_UINT,
+             .isYuv = false,
+             .horizontalAlignmentPixels = 1,
+             .verticalAlignmentPixels = 1,
+             .planes =
+                 {
+                     {
+                         .horizontalSubsampling = 1,
+                         .verticalSubsampling = 1,
+                         .bytesPerPixel = 8,
+                     },
+                 },
+         }},
     });
     return *kFormatInfoMap;
 }
diff --git a/src/gfxstream/guest/platform/include/VirtGpu.h b/src/gfxstream/guest/platform/include/VirtGpu.h
index 8f7ee182fc4..bedebb9241d 100644
--- a/src/gfxstream/guest/platform/include/VirtGpu.h
+++ b/src/gfxstream/guest/platform/include/VirtGpu.h
@@ -14,6 +14,8 @@
 #define VIRGL_FORMAT_B8G8R8A8_UNORM 1
 #define VIRGL_FORMAT_B5G6R5_UNORM 7
 #define VIRGL_FORMAT_R10G10B10A2_UNORM 8
+#define VIRGL_FORMAT_Z32_FLOAT_S8X24_UINT 126
+
 #define VIRGL_FORMAT_R8_UNORM 64
 #define VIRGL_FORMAT_R8G8B8_UNORM 66
 #define VIRGL_FORMAT_R8G8B8A8_UNORM 67
diff --git a/src/gfxstream/guest/vulkan_enc/ResourceTracker.cpp b/src/gfxstream/guest/vulkan_enc/ResourceTracker.cpp
index 4e0dc43c892..b3f8710d244 100644
--- a/src/gfxstream/guest/vulkan_enc/ResourceTracker.cpp
+++ b/src/gfxstream/guest/vulkan_enc/ResourceTracker.cpp
@@ -148,13 +148,13 @@ struct StagingInfo {
     }
 
     ~StagingInfo() {
-        for (auto stream : streams) {
-            delete stream;
-        }
-
         for (auto encoder : encoders) {
             delete encoder;
         }
+
+        for (auto stream : streams) {
+            delete stream;
+        }
     }
 
     void pushStaging(CommandBufferStagingStream* stream, VkEncoder* encoder) {
